lib/pure/asyncmacro.nim

```diff
@@ -377,35 +378,38 @@ proc asyncSingleProc(prc: NimNode): NimNode {.compileTime.} =
 
     var closureIterator = newProc(iteratorNameSym, [newIdentNode("FutureBase")],
                                   procBody, nnkIteratorDef)
-    closureIterator[4] = newNimNode(nnkPragma, prc[6]).add(newIdentNode("closure"))
+    closureIterator.pragma = copyNimTree(prc.pragma)
+    closureIterator.addPragma(newIdentNode("closure"))
+    closureIterator.addPragma(newIdentNode("gcsafe"))
     outerProcBody.add(closureIterator)
 
     # -> createCb(retFuture)
     #var cbName = newIdentNode("cb")
     var procCb = getAst createCb(retFutureSym, iteratorNameSym,
-                         newStrLitNode(prc[0].getName),
+                         newStrLitNode(prcName),
                          createFutureVarCompletions(futureVarIdents, nil))
     outerProcBody.add procCb
 
     # -> return retFuture
-    outerProcBody.add newNimNode(nnkReturnStmt, prc[6][prc[6].len-1]).add(retFutureSym)
+    outerProcBody.add newNimNode(nnkReturnStmt, prc.body[^1]).add(retFutureSym)
 
   result = prc
 
-  # Remove the 'async' pragma.
-  for i in 0 .. <result[4].len:
-    if result[4][i].kind == nnkIdent and result[4][i].ident == !"async":
-      result[4].del(i)
-  result[4] = newEmptyNode()
+  # If prc is proc declaration after forward declaration and pragma node kind is
+  # nnkPragma, Nim will complain that pragmas are not allowed even if the node
+  # has no children. We're replacing empty nnkPragma node with nnkEmpty here.
```

Comment:
Please, see https://github.com/nim-lang/Nim/blob/devel/compiler/semstmts.nim#L1327. Nim expects only `nkEmpty` node as pragma node. In our case we have `nkPragma` node which appeared because `async` pragma was once added and then removed. Even though our `nkPragma` node doesn't actually contain any pragmas Nim will complain. Dunno, maybe Nim should check for pragma node len instead of `nkEmpty` kind, but that fix looked more dangerous to me. @Araq, should I fix it in Nim instead?